### ISSUE TITLE
Naming of models in the new @binding API

### ISSUE THREAD

**Issue by hhaensel** [GenieFramework/Stipple.jl#143]
@essenciary 
The new API is really cool stuff. Here's my idea to make it more robust and performant (still not breaking 🤣 )

We introduce an abstract subtype of ReactiveModels. All model-specific definitions like js_watch would always be defined for the abstract type.

It could be along the following lines:
```julia

function Stipple.root(::M)::String where {M<:ReactiveModel}
  root(M)
end

# remove extra chars from the `gensym()` call and convert "." to ","
# if we define the real model name to be the abstract model name prefixed with `'#'`
# the `vm()` will be identical for real and abstract type
function Stipple.root(::Type{M})::String where {M<:ReactiveModel}
  replace(replace(string(M), "." => "_", r"var\"#*([^#]+).*\"" => s"\1"), r"[^0-9a-zA-Z_-]+" => "")
end

# ------------------------------------------------------------
abstract type HHModel <: ReactiveModel end

# all js definitions go with the abstract model type
js_watch(::HHModel) = "console.log('Hi')"

# the real model type is defined as subtype of the abstract type
@reactive! mutable struct var"#HHModel" <: HHModel
    a::R{Int} = 1
    b::R{Float64} = 2
end

model = init(var"#HHModel")

vm(HHModel)
# "HHModel"

vm(var"#HHModel")
# "HHModel"

vm(model)
# "HHModel"

js_watch(model)
# "console.log('Hi')"

@reactive! mutable struct var"#HHModel" <: HHModel
    a::R{Int} = 1
    b::R{Float64} = 2
end

# after redefinition we need to make sure deps and routes are replaced
# for this to work we need to modify `stipple_deps()` to register the route under Symbol(vm(m)) instead of Symbol(m)
delete!(Stipple.DEPS, filter(x -> x isa Type && x <: HHModel, keys(Stipple.DEPS)))
Genie.Router.delete!(Symbol(vm(HHModel)))

# So the `@binding` macro should do something like

@reactive! mutable struct var"#HHModel" <: HHModel
    a::R{Int} = 1
    b::R{Float64} = 2
    c::R{Float64} = 3
end

delete!(Stipple.DEPS, filter(x -> x isa Type && x <: HHModel, keys(Stipple.DEPS)))
Genie.Router.delete!(Symbol(vm(HHModel)))

# -----------------------  testing  --------------------------

Stipple.DEPS[:test] = () -> ["Just for testing that we don't delete too much..."]

route("/") do 
    page(model, "Hello")
end

up()

Genie.Router.routes()
Stipple.DEPS
```
Any js functions would be preserved without redefinition and inference. That should be much more performant.
What do you think?

**Comment by hhaensel** [comment]
Well, if two models have the same vm name, they share the same route.
It's highly improbable though, if we keep the module names in the vm-string. It only happens, if there are some ugly characters in the type name.

**Comment by essenciary** [comment]
I have addressed the "same model name" issue already by including the name of the module that contains the model in the model name. This combination has to be unique because it's enforced by the compiler.

**Comment by hhaensel** [comment]
I put all definitions in a macro. It's quite compact
```julia
using Stipple
import Stipple: js_watch

Stipple.root(::M) where M<:ReactiveModel = Stipple.root(M)
Stipple.root(::Type{M}) where M<:ReactiveModel = replace(replace(string(M), "." => "_", r"var\"#*([^#]+).*\"" => s"\1"), r"[^0-9a-zA-Z_]+" => "")

initmodel(m; kwargs...) = init(isabstracttype(m) ? eval(Symbol(m, '!')) : m; kwargs...)

macro newmodel(modelname, expr)
    modelconst = Symbol(modelname, '!')
    realname = gensym(modelname)

    esc(quote
        abstract type $modelname <: ReactiveModel end
        
        $modelconst = @reactive mutable struct $realname <: $modelname
            $(expr.args...)
        end

        delete!.(Ref(Stipple.DEPS), filter(x -> x isa Type && x <: $modelname, keys(Stipple.DEPS)))
        Genie.Router.delete!(Symbol(vm($modelname)))

        $modelconst
    end)
end

#  -----------  testing  -------------
@newmodel HHModel begin
    a::Int = 1
    b::Float64 = 2
end

hhmodel = initmodel(HHModel)

vm(hhmodel)
# HHModel

vm(HHModel)
# HHModel

js_watch(::M) where M <: HHModel = "HH"
js_watch(hhmodel)

# HH
```

**Comment by hhaensel** [comment]
For the REACTIVE_STORAGE, I'd follow the same concept but rather store expressions for the model types and not for the modules. That way it is easy to have more than one model in one module.

```julia
using OrderedCollections
REACTIVE_STORAGE = LittleDict{Type{<:ReactiveModel}, LittleDict{Symbol, Expr}}()

function Base.push!(rs::LittleDict, m::Union{ReactiveModel, Type{<:ReactiveModel}}, expr::Expr)
    mkey = modeltype(m)
    f = expr
    while ! isa(f, Symbol)
        f = f.args[1]
    end 
    haskey(rs, mkey) ? push!(rs[mkey], f => expr) : push!(rs, mkey => LittleDict(f => expr))
end

expr = :(a::R{Integer} = 1 + 3)
push!(REACTIVE_STORAGE, HHModel, :(a::R{Integer} = 1 + 3))
push!(REACTIVE_STORAGE, HHModel, :(b::R{Integer} = 2))
push!(REACTIVE_STORAGE, HHModel, :(c::R{Float64} = 3))

REACTIVE_STORAGE[HHModel]

macro model_from_storage(modelname)
    output = Core.eval(__module__, :(
        values(REACTIVE_STORAGE[$modelname])
    ))

    modelstring = string(modelname, "_")
    modelconst = Symbol(modelstring)
    realname = gensym(modelname)

    esc(quote

        abstract type $modelname <: ReactiveModel end

        $modelconst = @reactive mutable struct $realname <: $modelname
            $(output...)
        end

        for p in Stipple.DEPS
            p[1] isa Type && p[1] <: HHModel && delete!(Stipple.DEPS, p[1])
        end

        # oldname != "" && Genie.Router.delete!(oldname)
        for r in Genie.Router.routes()
            endswith(string(r.name), r"var\"##HHModel#\d+\"") && Genie.Router.delete!(r.name)
        end

        $modelconst
    end)
end

@model_from_storage HHModel
```
It shouldn't be difficult to define a default model for the module.
That way we would have the best of all worlds:

- hidden models for users who don't want to deal with models
- overwritable models
- heritage of js_methods without redefinition
- automatic cleaning of routes and DEPS

**Comment by hhaensel** [comment]
js_methods could be also defined by a macro
```julia
macro js_methods(modelname, methods)
    quote
        Stipple.js_methods(::M) where M <: $modelname = $methods
    end
end

@js_methods HHModel """
    increaseme: function() {this.a += 1}
    """

js_methods(model)
# "increaseme: function() {this.a += 1}\n"
```

**Comment by essenciary** [comment]
I think we need to keep the storage by module (so it could be by module and type). Because the macros are executed in the context of the module, we can easily check and retrieve what is defined in the context of that module. Also, this guarantees that there are no type name conflicts between different types defined in different modules. 

However, multiple models per module looks like premature optimisation at this point. I would prefer to roll out the implicit model API ASAP, with one model per module, and get feedback from users. If multiple implicit models will be a common requirement (though I haven't seen this use case with explicit models so far) we can add it later. Otherwise less code and complexity is better -- there's less stuff to debug, document, and maintain. 

Re subtypes and `js_methods` that sounds great. I'll leave it up to you as I don't have a close knowledge of that codebase nor do I use it on a regular basis. Just please don't break the implicit models API that I'm rolling out (so again I'd prefer to leave `REACTIVE_STORAGE` for now as that's breaking and I want to release the API ASAP).

**Comment by hhaensel** [comment]
@essenciary I want to revive this topic as it is really annoying that all modeltype-based definitions get lost after redefining the model. I'v put some work into it and it just works. It could be easily implemented in your ReactiveTools as well.
Below is the code that I ended up with.
```julia
Stipple.root(::M) where M<:ReactiveModel = Stipple.root(M)
Stipple.root(::Type{M}) where M<:ReactiveModel = string(rootname(M))

rootname(x) = Symbol(replace(replace(string(x), "." => "_", r"var\"#*([^#]+).*\"" => s"\1"), r"[^0-9a-zA-Z_]+" => ""))

function Stipple.stipple_deps(m::Type{M}, vue_app_name, debounce, core_theme, endpoint, transport)::Function where {M<:ReactiveModel}
    @info "here"
    () -> begin
      if ! Genie.Assets.external_assets(Stipple.assets_config)
        if ! Genie.Router.isroute(rootname(M))
          Genie.Router.route(Genie.Assets.asset_route(Stipple.assets_config, :js, file = endpoint), named = rootname(M)) do
            Stipple.Elements.vue_integration(m; vue_app_name, debounce, core_theme, transport) |> Genie.Renderer.Js.js
          end
        end
      end
  
      [
        if ! Genie.Assets.external_assets(Stipple.assets_config)
          Genie.Renderer.Html.script(src = Genie.Assets.asset_path(Stipple.assets_config, :js, file = vue_app_name), defer = true)
        else
          Genie.Renderer.Html.script([
            (Stipple.Elements.vue_integration(m; vue_app_name, core_theme, debounce) |> Genie.Renderer.Js.js).body |> String
          ])
        end
      ]
    end
  end

function initmodel(m::Type{<:ReactiveModel}; kwargs...)
    model = isabstracttype(m) ? Core.eval(Base.parentmodule(m), Symbol(Base.nameof(m), "!")) : m
    Base.parentmodule(m), Symbol(Base.nameof(m), "!")
    Base.invokelatest(init, model; kwargs...)
end

macro model(modelname, expr)
    modelconst = Symbol(modelname, '!')

    esc(quote
        abstract type $modelname <: ReactiveModel end
        
        @reactive! mutable struct $modelconst <: $modelname
            $(expr.args...)
        end

        delete!.(Ref(Stipple.DEPS), filter(x -> x isa Type && x <: $modelname, keys(Stipple.DEPS)))
        Genie.Router.delete!(rootname($modelname))
        
        $modelconst
    end)
end
```
```julia
@model HHModel begin
    a::Int = 1
    b::Float64 = 2
end
```
Should I prepare a PR?

**Comment by hhaensel** [comment]
Note, I defined `rootname()` to strip the number from the gensym name in order to have always the same name of the js file.
And I just added a correction above that fixes backwards compatibility.

**Comment by hhaensel** [comment]
Just edited the code above...

**Comment by hhaensel** [comment]
```julia
@reactive! mutual struct HHModel <: ReactiveModel
    a::R{Int} = 1
    b::R{Float64} = 2
end
```
is replaced with
```julia
@model HHModel begin
    a::R{Int} = 1
    b::R{Float64} = 2
end
```

**Comment by hhaensel** [comment]
Thought a  bit about naming. `@model` creates a model instance, therefore I renamed my macro to `@modeltype`

**Comment by hhaensel** [comment]
Reading through ReactivityTools I think a better naming would be `@type` instead of `@modeltype`

**Comment by hhaensel** [comment]
@essenciary 
I have pushed a working version now and it's ready for testing.

After refactoring the ReactiveTools I still propose the following changes:

- `@reactive!` works as is.

- `@type begin ... end` will be the recommended replacement as discussed above.

- `@type` accepts a new syntax which is closer to the one you used in ReactiveTools, e.g.
   ```julia
    @type begin
        a = "Hello"
        b::AbstractString = "World"
        n = 10
        m::Real = 11, READONLY
        readonly = "hi", READONLY, NON_REACTIVE
        private = "you don't see me", PRIVATE, NON_REACTIVE
    end
   ```

**Comment by hhaensel** [comment]
BTW, we currently don't support listeners for non_reactive fields, but they could be triggered manually. Should we change that?

**Comment by essenciary** [comment]
Cool! so the `@in` `@out` etc still work as previous? This is quite important for my use case as I want users to be able to tag variables where they appear in their code, vs explicitly declaring types. With the ReactiveTools API the whole idea of a type/model is hidden. 

Thinking about this, there is really no trace of types in the new API, so a `@type` macro would leave people wondering. Maybe `@vars` or something makes more sense? 

Re the `@type` example that's practically the `@reactive` declaration, minus a few words. The benefit or using it is that users don't chose the name of the type and the name of the model does not change when the fields change?

**Comment by essenciary** [comment]
> BTW, we currently don't support listeners for non_reactive fields, but they could be triggered manually. Should we change that?

I don't understand the implications of this. If the users don't want to make the fields reactive, why have listeners?

**Comment by hhaensel** [comment]
The only possible application I see is parameters that are typically changed by the front end and not by the backend.
But I think we should be clear by design. We allow that the value is changed from the front end, so if someone executes the correct js code, the value will be changed. If we don't want that we should exclude that.
If we don't install listeners but want to keep the possiibility of updates, we should provide a js function that does trigger such change, e.g. `push(model, fieldname, value)` in analogy to `push!(model, fieldname => value)`. I do use the latter one to change client_data.

I just tested installing listeners, that's easy and works. So now we have the choice ...

**Comment by hhaensel** [comment]
I thought about it, I tend to have no listeners for non_reactives. But I would offer a js-push method.
It makes the setup more understandable:
- Reactive: with listeners
- Non-Reactive: without listeners
- PUBLIC: pushing from both sides allowed
- READONLY: no pushing from fron-end
- PRIVATE: not shared to frontend

We could abandon NO_WATCHER in that case. I don't see a use case where we have no backend watcher. Switching off frontend watcher is done by READONLY.

**Comment by hhaensel** [comment]
We discussed this quite a while ago, but we said we leave it as is. But during all that time I never had a use case

**Comment by hhaensel** [comment]
> Cool! so the `@in` `@out` etc still work as previous?

Yip! Even better, it's also possible to just import Stipple and ReactiveTools and it will work as well.

> Maybe `@vars` or something makes more sense?
Sounds good for the ReactiveTools. I'd still favour `@type` for people who want to declare their own models.

> Re the `@type` example that's practically the `@reactive` declaration, minus a few words.

Exactly.

> The benefit or using it is that users don't chose the name of the type and the name of the model does not change when the fields change?

The benefits are
- the model doesn't change its name
- all type-specific function declarations will keep their validity without redefining
- old syntax is supported
- new syntax brings
    - no type declarations needed if the type is well-defined by the initial value
    - reactive type declaration without surrounding `R{}`. This is nice for people that use `import` rather than `using` because the parse_expression will check this and prefix with `Stipple.`otherwise.
    - non_reactive types are declared in a very similar way and types are also automatically attached

But this still needs to be written ...

BTW, I added a js push function to do `model.push(field)` which works for both, reactive and non-reactive types as long as they are `PUBLIC`

**Comment by essenciary** [comment]
Amazing features! Let me know how I can help to release them.

**Comment by hhaensel** [comment]
> Amazing features! Let me know how I can help to release them.

Test it 😄
There is `@vars` and `@add_vars` both supporting old and new syntax.  Old Syntax is chosen as soon as an explicit Reactive Type is detected, otherwise new Syntax is used.

```julia
@vars begin
    s1 = "Hello"
    s2 = "World", NON_REACTIVE, PRIVATE
    n::Int = 12, JSFUNCTION
end

@add_vars begin
    s1 = "Hi"
    s2 = "there", NON_REACTIVE, READONLY
    m::Int = 13, PRIVATE
end
```
Later definitions overwrite (as in `merge()`)

**Comment by hhaensel** [comment]
Solved by #151
